### PR TITLE
bpo-25988: Do not expose abstract collection classes in the collections module.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue25988

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
I think it makes sense to ping pip folks about the test failure. @pfmoore @dstufft not something urgent, since 3.8 is still far away, but it would be great if `pip` didn't use the deprecated `collections` imports like `collections.Mapping` and switched to `collections.abc.Mapping` instead, thanks!.

**Comment by dstufft** [comment]
This is actually in urllib3, we just bundle urllib3, it looks fixed in urllib3 and in pip and we just need to update CPython's bundled copy of pip.

**Comment by ilevkivskyi** [comment]
> we just need to update CPython's bundled copy of pip.

Great, thanks for looking into this!

**Comment by dstufft** [comment]
The above has been merged, you should be able to rebase this PR.

**Comment by tirkarthi** [comment]
This might cause `make doctest` to fail since some of the build dependencies might be incompatible https://bugs.python.org/issue35109

cc: @JulienPalard

**Review by rhettinger — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
Tests still are failing because of incompatibility in `html5lib` bundled with `pip`.
```
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/tmpcbwhlouc/pip-18.1-py2.py3-none-any.whl/pip/_vendor/html5lib/_trie/_base.py", line 3, in <module>
ImportError: cannot import name 'Mapping' from 'collections' (/home/travis/build/python/cpython/Lib/collections/__init__.py)
```

Note also syntax warnings produced by `requests`.
```
/tmp/tmpocpc284j/pip-18.1-py2.py3-none-any.whl/pip/_vendor/requests/status_codes.py:18: SyntaxWarning: invalid escape sequence \o
```

**Comment by tirkarthi** [comment]
Requests also seem to be fixed : https://github.com/requests/requests/pull/4728

So I guess pip's vendor folder needs to be updated for both libraries?

**Review by rhettinger — APPROVED** [review]

**Comment by rhettinger** [comment]
FWIW, it would be reasonable to delay this be one additional point release.

Also, if there are known major projects that still do the old imports, we should submit bug reports to those projects (I've already done some of these).

**Comment by tirkarthi** [comment]
> Also, if there are known major projects that still do the old imports, 

Our docs build has this problem with a dependency that had this issue and now it's fixed upstream. I think it's better to do it now that gives more time and will help users and libraries who test their code with alpha and beta candidates if they have these imports.

**Comment by tirkarthi** [comment]
@dstufft @pradyunsg https://bugs.python.org/issue35807 updates pip to 19.0 but I still see the vendored copy using html5lib that has this warning. Does pip update vendored version copy only after a release from html5lib? Should I file an issue in Pip's repo for this?

**Comment by pradyunsg** [comment]
(sorry for the delay in responding; swamped in GitHub notifications currently)

> Does pip update vendored version copy only after a release from html5lib?

Yep. pip downloads and vendors only versions released on PyPI.

> Should I file an issue in Pip's repo for this?

This isn't _necessary_ since we basically do a round of vendoring updates before a release which should cover this, but yeah, it won't hurt either. Please feel free to.

**Comment by serhiy-storchaka** [comment]
pip 19.0.3 still is not compatible with this PR.

**Comment by pradyunsg** [comment]
Yea, there hasn't been an update for html5lib yet. :(

I pinged the author on #pypa-dev on Freenode today about this.

**Comment by methane** [comment]
If we postpone the removal, can we change warning class from DeprecationWarning to ImportWarning to nosy more people?

Maybe, many Python developers doesn't care about DeprecationWarning after Python 3.2.

**Comment by njsmith** [comment]
My impression is that there's a *lot* of code out there that uses the ABCs directly from `collections`. It might make sense to break this for the alpha to encourage people to upgrade and get a sense of how widespread the breakage is, but I think there's a good chance that ultimately we'll want to defer it to 3.9.

**Comment by asvetlov** [comment]
I see several libraries that still import abstract classes from `collections` directly.
Postponing the removal make sense maybe but I have no idea how it can help with fixing these libraries if they ignore deprecation for many years.
Is `ImportWrning` more visible than `DeprecationWarning`?

**Comment by njsmith** [comment]
The deprecation warming was only added in 3.7, so it hasn't even been a year yet. Probably there will be projects that ignore it for "many years", but so far they haven't far a chance to do that :-)

The only motivation here is tidiness, right? This isn't blocking some important change, or creating undue maintenance burden, right?

**Comment by Carreau** [comment]
> ... Python developers doesn't care about DeprecationWarning...
> ... I see several libraries that still import abstract classes from collections directly....
> ... My impression is that there's a lot of code out there that uses the ABCs directly from collections...
> ... Probably there will be projects that ignore it for "many years"...


I don't believe is that's we "don't care", I – at least – do care, but the messaging does not in my humble opinion,  give enough informations on the spot  for a quick actionable patch to my codebases.  I believe that some extra care could go a long way toward nudging people to update, ant not using a slick (removed in XXX), but a carrot.

First, I must say that the DeprecationWarnings have become way better in the last Couple of years and among other @serhiy-storchak have done a fantastic job cleaning  many of them up. I believe we could do even better. 

In this specific example , deprecation message was added in 3.7 (as Nataniel point out),  message mention removal in 3.8. When encountering it I'm likely not going to go hunt  for more informations. If I'm in a good mood I'll look in the source for comments, (which mention both 3.6 and 3.7.), and if I'm feeling adventurous the online docs (which mention 3.3, but as a "Change" not a "Deprecation"). And here I'm lucky I did not have to git-blame to find the change – which used to often be the case. 

So now, as a library author I'm unsure when I can rely in things being in `collections.abc` and when to do conditional import depending on Python version; I'm thus unlikely to go patch my code. 

A "please import from `collections.abc` on Python 3.3 and above" in the DeprecationWarning would have given me a way clearer idea that yeah, I could have replaced  `collections` by `collections.abc` and it would have worked on all the Python versions I'm compatible with and nudge me to patch my code if not now – or at least soon.

I understand that when working on Core Python you start to get a feeling of when things change and what to do about it, but for users not regularly following, it can be quite tough to know what to do, and what are the implications. I hope that some of the deprecation messages could be shifted to give a couple of more actionable actions to users.

I'm of course happy to send PRs – and already send a few to improve the deprecation messages through the years.

**Comment by serhiy-storchaka** [comment]
Seems that one release of active deprecation is not enough for the change which affects so much code. I am fine with deferring the removal to 3.9 and changing the warning message. @Carreau, do you mind to create a PR?

**Comment by Carreau** [comment]
> @Carreau, do you mind to create a PR?

Sure thing, done as #13409.

**Comment by dhomeier** [comment]
I'd agree that this was not very prominently flagged before, and for casual developers it is not straightforward to see why `ABCMeta`-derived classes like `Awaitable` or `Iterable` raise the warning, while things like `OrderedDict` are still directly imported from `collections`.

**Comment by vstinner** [comment]
I wasn't aware of this PR. I proposed the same idea: "collections: remove deprecated aliases to ABC classes"

* https://bugs.python.org/issue37324
* https://github.com/python/cpython/pull/14171

... and I'm also blocked by html5lib: https://github.com/html5lib/html5lib-python/issues/419

**Comment by pradyunsg** [comment]
pip now holds a patch for html5lib and the next release of pip (due this month), will have that patch.

**Comment by yan12125** [comment]
New pip versions are released and the pip wheel inside CPython is updated to 19.2.1 in 5380def8269b24a8a3bc46396373a1dc91b1dd1a. I believe this pull request can be revisited now.

**Comment by pradyunsg** [comment]
@yan12125 is faster than me. :P

File: Doc/whatsnew/3.8.rst
Comment (yan12125) [inline Doc/whatsnew/3.8.rst]: I think this note can be removed as there is already a note in `Doc/whatsnew/3.9.rst`?

**Comment by serhiy-storchaka** [comment]
Seems [Jinja2](https://github.com/pallets/jinja) is only a stopper. The issue was [fixed](https://github.com/pallets/jinja/pull/867) in 2.10.2 more than a year ago, but this version is not released yet.

**Comment by pradyunsg** [comment]
Where is jinja2 used?

**Comment by serhiy-storchaka** [comment]
It is used in `sphinx-build`.

**Comment by pradyunsg** [comment]
Ah. So that explains the documentation build failing. Thanks for the super quick response to a trivial question!

**Comment by yan12125** [comment]
I gave it a try - the jinja issue is the last blocker. If I use an unreleased commit of jinja [1], tests are green [2].

[1] https://github.com/yan12125/cpython/commit/e6641707fb87774bf6359d8a584047cc94cc5b68#diff-c8dee9c3c068ad3447bbf459950a35b5
[2] https://travis-ci.org/yan12125/cpython/builds/567688247

**Comment by vstinner** [comment]
> The issue was fixed in 2.10.2 more than a year ago, but this version is not released yet.

That's surprising. Why was a tag created in Git with no public release on PyPI?

Jinja2 fix for collections.abc: https://github.com/pallets/jinja/commit/593ee1e481d9534edd65928ac6bae5cfbd2eec14

**Comment by yan12125** [comment]
Seems 2.10.2 is not tagged yet (or it was tagged but later removed): https://github.com/pallets/jinja/tags

**Comment by anowlcalledjosh** [comment]
There was an issue filed against Jinja2 in April asking when the fix would be included in a release, but the maintainers aren't saying much: pallets/jinja#973

**Comment by yan12125** [comment]
A new version of Jinja is out - http://jinja.palletsprojects.com/en/2.10.x/changelog/#version-2-10-2. This pull request can go forward now.

And the git conflict seems trivial to fix :) https://github.com/python/cpython/commit/c4106af38bbcb180725fe0d9478e6a11f7a5e7b9

**Comment by pradyunsg** [comment]
The new jinja release has been made! We're unblocked on this! :)

**Comment by yan12125** [comment]
I gave it another try - with this PR rebased [1] and Jinja 2.10.3 installed, tests are still green [2].

[1] https://github.com/yan12125/cpython/commits/collections-abc
[2] https://travis-ci.org/yan12125/cpython/builds/593814893

**Comment by pradyunsg** [comment]
Yay! We managed to do this in under an year! ^>^

**Pull request merged**
